File: game/match_contract.lua
local arenas = require("data.arenas")
local formations = require("data.formations")
local players = require("data.players")
local tactics = require("data.tactics")
local teams = require("data.teams")

---@alias MatchWinner "home"|"away"|"draw"

---@class ProductMatchRequest
---@field home_team_id string
---@field away_team_id string
---@field home_starter_ids string[]
---@field formation_id string
---@field tactic_id string
---@field arena_id string
---@field show_onboarding boolean
---@field combat_enabled boolean
---@field seed integer?

---@class TeamResultStats
---@field shots integer?
---@field possession number?
---@field saves integer?
---@field pass_completion number?

---@class ProductMatchResult
---@field home_team_id string
---@field away_team_id string
---@field home_name string
---@field away_name string
---@field home_score integer
---@field away_score integer
---@field winner MatchWinner
---@field mvp_player_id string?
---@field mvp_summary string?
---@field home_stats TeamResultStats
---@field away_stats TeamResultStats
---@field seed integer?

---@class ProductMatchRequestOptions
---@field home_team_id string
---@field away_team_id string
---@field home_starter_ids string[]
---@field formation_id string
---@field tactic_id string
---@field arena_id string?
---@field show_onboarding boolean?
---@field combat_enabled boolean?
---@field seed integer?

---@class ProductMatchResultOptions
---@field home_team_id string
---@field away_team_id string
---@field home_score integer
---@field away_score integer
---@field mvp_player_id string?
---@field mvp_summary string?
---@field home_stats TeamResultStats?
---@field away_stats TeamResultStats?
---@field seed integer?

---@class MatchContractModule
local contract = {}

---@return table<string, PlayerData>
local function players_by_id()
    local result = {}
    for _, player in ipairs(players) do
        result[player.id] = player
    end
    return result
end

---@param ids string[]
---@return string[]
local function copy_ids(ids)
    local result = {}
    for i, id in ipairs(ids) do
        result[i] = id
    end
    return result
end

---@param team TeamData
---@return table<string, boolean>
local function eligible_ids(team)
    local result = {}
    local squad = team.squad or team.roster
    for _, id in ipairs(squad) do
        result[id] = true
    end
    return result
end

---@param ids string[]
---@param team TeamData
---@return boolean?, string?
function contract.validate_starters(ids, team)
    if #ids ~= 5 then
        return nil, "a team sheet needs exactly five starters"
    end

    local pool = players_by_id()
    local eligible = eligible_ids(team)
    local seen = {}
    local keepers = 0
    for _, id in ipairs(ids) do
        if seen[id] then
            return nil, "starter ids must be unique"
        end
        seen[id] = true
        if not eligible[id] then
            return nil, "player is not eligible for " .. team.name .. ": " .. id
        end
        local player = pool[id]
        if not player then
            return nil, "unknown player id: " .. id
        end
        if player.position == "keeper" then
            keepers = keepers + 1
        end
    end
    if keepers ~= 1 then
        return nil, "a team sheet needs exactly one keeper"
    end
    return true
end

---@param opts ProductMatchRequestOptions
---@return ProductMatchRequest?, string?
function contract.new_request(opts)
    local home = teams[opts.home_team_id]
    if not home then
        return nil, "unknown home team: " .. tostring(opts.home_team_id)
    end
    if not teams[opts.away_team_id] then
        return nil, "unknown away team: " .. tostring(opts.away_team_id)
    end
    if not formations[opts.formation_id] then
        return nil, "unknown formation: " .. tostring(opts.formation_id)
    end
    if not tactics[opts.tactic_id] then
        return nil, "unknown tactic: " .. tostring(opts.tactic_id)
    end
    local arena_id = opts.arena_id or "helios_crown"
    if not arenas[arena_id] then
        return nil, "unknown arena: " .. tostring(arena_id)
    end
    local ok, err = contract.validate_starters(opts.home_starter_ids, home)
    if not ok then
        return nil, err
    end

    return {
        home_team_id = opts.home_team_id,
        away_team_id = opts.away_team_id,
        home_starter_ids = copy_ids(opts.home_starter_ids),
        formation_id = opts.formation_id,
        tactic_id = opts.tactic_id,
        arena_id = arena_id,
        show_onboarding = opts.show_onboarding == true,
        combat_enabled = opts.combat_enabled == true,
        seed = opts.seed,
    }
end

---@param value number?
---@return boolean
local function valid_ratio(value)
    return value == nil or (value >= 0 and value <= 1)
end

---@param value integer?
---@return boolean
local function valid_count(value)
    return value == nil or (value >= 0 and value == math.floor(value))
end

---@param stats TeamResultStats
---@return boolean?, string?
local function validate_stats(stats)
    if not valid_count(stats.shots) then
        return nil, "shots must be a non-negative integer"
    end
    if not valid_count(stats.saves) then
        return nil, "saves must be a non-negative integer"
    end
    if not valid_ratio(stats.possession) then
        return nil, "possession must be between zero and one"
    end
    if not valid_ratio(stats.pass_completion) then
        return nil, "pass completion must be between zero and one"
    end
    return true
end

---@param opts ProductMatchResultOptions
---@return ProductMatchResult?, string?
function contract.new_result(opts)
    local home = teams[opts.home_team_id]
    local away = teams[opts.away_team_id]
    if not home or not away then
        return nil, "match result references an unknown team"
    end
    if
        not valid_count(opts.home_score)
        or opts.home_score == nil
        or not valid_count(opts.away_score)
        or opts.away_score == nil
    then
        return nil, "scores must be non-negative integers"
    end

    local home_stats = opts.home_stats or {}
    local away_stats = opts.away_stats or {}
    local ok, err = validate_stats(home_stats)
    if not ok then
        return nil, err
    end
    ok, err = validate_stats(away_stats)
    if not ok then
        return nil, err
    end

    local winner = "draw"
    if opts.home_score > opts.away_score then
        winner = "home"
    elseif opts.away_score > opts.home_score then
        winner = "away"
    end

    return {
        home_team_id = home.id,
        away_team_id = away.id,
        home_name = home.name,
        away_name = away.name,
        home_score = opts.home_score,
        away_score = opts.away_score,
        winner = winner,
        mvp_player_id = opts.mvp_player_id,
        mvp_summary = opts.mvp_summary,
        home_stats = home_stats,
        away_stats = away_stats,
        seed = opts.seed,
    }
end

return contract

File: ts/packages/app/src/match_contract.ts
// Arena/formation/player/tactic/team data is Rust-owned (ARCHITECTURE.md §4
// rule 6), so every function here takes a `MatchContractContent` bundle as its
// first explicit parameter instead of importing those tables.

import { type Result, err, ok } from "@gc/core";
import type { MatchContractContent, TeamData } from "./content.ts";

export type MatchWinner = "home" | "away" | "draw";

export interface ProductMatchRequest {
  readonly home_team_id: string;
  readonly away_team_id: string;
  readonly home_starter_ids: readonly string[];
  readonly formation_id: string;
  readonly tactic_id: string;
  readonly arena_id: string;
  readonly show_onboarding: boolean;
  readonly combat_enabled: boolean;
  readonly seed?: number;
}

export interface TeamResultStats {
  readonly shots?: number;
  readonly possession?: number;
  readonly saves?: number;
  readonly pass_completion?: number;
}

export interface ProductMatchResult {
  readonly home_team_id: string;
  readonly away_team_id: string;
  readonly home_name: string;
  readonly away_name: string;
  readonly home_score: number;
  readonly away_score: number;
  readonly winner: MatchWinner;
  readonly mvp_player_id?: string;
  readonly mvp_summary?: string;
  readonly home_stats: TeamResultStats;
  readonly away_stats: TeamResultStats;
  readonly seed?: number;
}

export interface ProductMatchRequestOptions {
  readonly home_team_id: string;
  readonly away_team_id: string;
  readonly home_starter_ids: readonly string[];
  readonly formation_id: string;
  readonly tactic_id: string;
  readonly arena_id?: string;
  readonly show_onboarding?: boolean;
  readonly combat_enabled?: boolean;
  readonly seed?: number;
}

export interface ProductMatchResultOptions {
  readonly home_team_id: string;
  readonly away_team_id: string;
  readonly home_score: number;
  readonly away_score: number;
  readonly mvp_player_id?: string;
  readonly mvp_summary?: string;
  readonly home_stats?: TeamResultStats;
  readonly away_stats?: TeamResultStats;
  readonly seed?: number;
}

function eligibleIds(team: TeamData): ReadonlySet<string> {
  return new Set(team.squad ?? team.roster);
}

export function validateStarters(
  content: Pick<MatchContractContent, "players">,
  ids: readonly string[],
  team: TeamData,
): Result<true, string> {
  if (ids.length !== 5) {
    return err("a team sheet needs exactly five starters");
  }

  const eligible = eligibleIds(team);
  const seen = new Set<string>();
  let keepers = 0;
  for (const id of ids) {
    if (seen.has(id)) {
      return err("starter ids must be unique");
    }
    seen.add(id);
    if (!eligible.has(id)) {
      return err(`player is not eligible for ${team.name}: ${id}`);
    }
    const player = content.players[id];
    if (!player) {
      return err(`unknown player id: ${id}`);
    }
    if (player.position === "keeper") {
      keepers += 1;
    }
  }
  if (keepers !== 1) {
    return err("a team sheet needs exactly one keeper");
  }
  return ok(true);
}

export function newRequest(
  content: MatchContractContent,
  opts: ProductMatchRequestOptions,
): Result<ProductMatchRequest, string> {
  const home = content.teams[opts.home_team_id];
  if (!home) {
    return err(`unknown home team: ${String(opts.home_team_id)}`);
  }
  if (!content.teams[opts.away_team_id]) {
    return err(`unknown away team: ${String(opts.away_team_id)}`);
  }
  if (!content.formations[opts.formation_id]) {
    return err(`unknown formation: ${String(opts.formation_id)}`);
  }
  if (!content.tactics[opts.tactic_id]) {
    return err(`unknown tactic: ${String(opts.tactic_id)}`);
  }
  const arenaId = opts.arena_id ?? "helios_crown";
  if (!content.arenas[arenaId]) {
    return err(`unknown arena: ${String(arenaId)}`);
  }
  const validated = validateStarters(content, opts.home_starter_ids, home);
  if (!validated.ok) {
    return validated;
  }

  return ok({
    home_team_id: opts.home_team_id,
    away_team_id: opts.away_team_id,
    home_starter_ids: [...opts.home_starter_ids],
    formation_id: opts.formation_id,
    tactic_id: opts.tactic_id,
    arena_id: arenaId,
    show_onboarding: opts.show_onboarding === true,
    combat_enabled: opts.combat_enabled === true,
    ...(opts.seed !== undefined ? { seed: opts.seed } : {}),
  });
}

function validRatio(value: number | undefined): boolean {
  return value === undefined || (value >= 0 && value <= 1);
}

function validCount(value: number | undefined): boolean {
  return value === undefined || (value >= 0 && value === Math.floor(value));
}

function validateStats(stats: TeamResultStats): Result<true, string> {
  if (!validCount(stats.shots)) {
    return err("shots must be a non-negative integer");
  }
  if (!validCount(stats.saves)) {
    return err("saves must be a non-negative integer");
  }
  if (!validRatio(stats.possession)) {
    return err("possession must be between zero and one");
  }
  if (!validRatio(stats.pass_completion)) {
    return err("pass completion must be between zero and one");
  }
  return ok(true);
}

export function newResult(
  content: { readonly teams: Readonly<Record<string, Pick<TeamData, "id" | "name">>> },
  opts: ProductMatchResultOptions,
): Result<ProductMatchResult, string> {
  const home = content.teams[opts.home_team_id];
  const away = content.teams[opts.away_team_id];
  if (!home || !away) {
    return err("match result references an unknown team");
  }
  if (
    !validCount(opts.home_score) ||
    opts.home_score === undefined ||
    !validCount(opts.away_score) ||
    opts.away_score === undefined
  ) {
    return err("scores must be non-negative integers");
  }

  const homeStats = opts.home_stats ?? {};
  const awayStats = opts.away_stats ?? {};
  const homeValidated = validateStats(homeStats);
  if (!homeValidated.ok) {
    return homeValidated;
  }
  const awayValidated = validateStats(awayStats);
  if (!awayValidated.ok) {
    return awayValidated;
  }

  let winner: MatchWinner = "draw";
  if (opts.home_score > opts.away_score) {
    winner = "home";
  } else if (opts.away_score > opts.home_score) {
    winner = "away";
  }

  return ok({
    home_team_id: home.id,
    away_team_id: away.id,
    home_name: home.name,
    away_name: away.name,
    home_score: opts.home_score,
    away_score: opts.away_score,
    winner,
    ...(opts.mvp_player_id !== undefined ? { mvp_player_id: opts.mvp_player_id } : {}),
    ...(opts.mvp_summary !== undefined ? { mvp_summary: opts.mvp_summary } : {}),
    home_stats: homeStats,
    away_stats: awayStats,
    ...(opts.seed !== undefined ? { seed: opts.seed } : {}),
  });
}

export const matchContract = { validateStarters, newRequest, newResult };
